### ISSUE TITLE
[installation fail] debian 9 (stretch, amd64), nodejs 10

### ISSUE THREAD

**Issue by GeraldWodni** [nodesource/distributions#747]
When I try to install node, the ancient debian version (`4.8.2~dfsg-1`) gets pulled instead of the shiny version 10.

The closest issue I found was #739, however I am running a 64 bit system.

`curl -sL https://deb.nodesource.com/setup_10.x | sudo -E bash -` yields:
```
## Installing the NodeSource Node.js 10.x repo...


## Populating apt-get cache...

+ apt-get update
Ign:1 http://deb.debian.org/debian stretch InRelease
Hit:2 https://deb.nodesource.com/node_10.x stretch InRelease
Get:3 http://deb.debian.org/debian sid InRelease [233 kB]
Hit:4 http://deb.debian.org/debian stretch Release          
Get:5 http://deb.debian.org/debian sid/main amd64 Packages.diff/Index [27.9 kB]
Get:6 http://deb.debian.org/debian sid/main Translation-en.diff/Index [27.9 kB]
Get:7 http://deb.debian.org/debian sid/main amd64 Contents (deb).diff/Index [28.0 kB]
Get:8 http://deb.debian.org/debian sid/main amd64 Packages 2018-10-31-1408.09.pdiff [19.3 kB]
Get:9 http://deb.debian.org/debian sid/main Translation-en 2018-10-31-1408.09.pdiff [1138 B]
Get:8 http://deb.debian.org/debian sid/main amd64 Packages 2018-10-31-1408.09.pdiff [19.3 kB]
Get:9 http://deb.debian.org/debian sid/main Translation-en 2018-10-31-1408.09.pdiff [1138 B]
Get:10 http://deb.debian.org/debian sid/main amd64 Contents (deb) 2018-10-31-1408.09.pdiff [1240 B]    
Get:10 http://deb.debian.org/debian sid/main amd64 Contents (deb) 2018-10-31-1408.09.pdiff [1240 B]             
Get:12 http://deb.debian.org/debian sid/main amd64 DEP-11 Metadata [3910 kB]                                    
Get:13 http://deb.debian.org/debian sid/main DEP-11 64x64 Icons [7906 kB]                                       
Fetched 12.2 MB in 7s (1526 kB/s)                                                                               
Reading package lists... Done

## Confirming "stretch" is supported...

+ curl -sLf -o /dev/null 'https://deb.nodesource.com/node_10.x/dists/stretch/Release'

## Adding the NodeSource signing key to your keyring...

+ curl -s https://deb.nodesource.com/gpgkey/nodesource.gpg.key | apt-key add -
OK

## Creating apt sources list file for the NodeSource Node.js 10.x repo...

+ echo 'deb https://deb.nodesource.com/node_10.x stretch main' > /etc/apt/sources.list.d/nodesource.list
+ echo 'deb-src https://deb.nodesource.com/node_10.x stretch main' >> /etc/apt/sources.list.d/nodesource.list

## Running `apt-get update` for you...

+ apt-get update
Ign:1 http://deb.debian.org/debian stretch InRelease
Hit:2 http://deb.debian.org/debian sid InRelease
Hit:3 http://deb.debian.org/debian stretch Release
Hit:4 https://deb.nodesource.com/node_10.x stretch InRelease
Reading package lists... Done

## Run `sudo apt-get install -y nodejs` to install Node.js 10.x and npm
## You may also need development tools to build native addons:
     sudo apt-get install gcc g++ make
## To install the Yarn package manager, run:
     curl -sL https://dl.yarnpkg.com/debian/pubkey.gpg | sudo apt-key add -
     echo "deb https://dl.yarnpkg.com/debian/ stable main" | sudo tee /etc/apt/sources.list.d/yarn.list
     sudo apt-get update && sudo apt-get install yarn
```

And installing it `sudo apt-get install -y nodejs` yields:
```
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following NEW packages will be installed:
  nodejs
0 upgraded, 1 newly installed, 0 to remove and 3 not upgraded.
Need to get 0 B/3440 kB of archives.
After this operation, 14.3 MB of additional disk space will be used.
Selecting previously unselected package nodejs.
(Reading database ... xxxxxxx files and directories currently installed.)
Preparing to unpack .../nodejs_4.8.2~dfsg-1_amd64.deb ...
Unpacking nodejs (4.8.2~dfsg-1) ...
Setting up nodejs (4.8.2~dfsg-1) ...
update-alternatives: using /usr/bin/nodejs to provide /usr/bin/js (js) in auto mode
Processing triggers for man-db (2.7.6.1-2) ...
```

In case it matters there is also no weird pinning going on, I just enabled sid (in `/etc/apt/preferences.d/limit-sid`):
```Package: *
Pin: release a=stable
Pin-Priority: 1000

Package: *
Pin: release a=sid
Pin-Priority: 150
```

This is the first time I'm running into this issue, all my other boxes are happy (but still on node8).

Any help appreciated

**Comment by chrislea** [comment]
The problem is that stretch is too old to run modern versions of V8, as V8 requires a reasonably new version of `libstdc++` to be able to compile and function. So we don't make packages for it. If you update to a current version like Jessie, you'll find supported packages.

**Comment by GeraldWodni** [comment]
@chrislea thanks for the answer, but stretch is the current stable version. Jessie is older. So my issue has nothing to do with a too old Debian.

**Comment by chrislea** [comment]
Ugh. Sorry. The sleep deprevation is real, apparently. What are the outputs of:
```
% cat /etc/apt/sources.list.d/nodesource.list
```
and also
```
% apt-cache policy nodejs
```
?

**Comment by GeraldWodni** [comment]
No worries ;)

% cat /etc/apt/sources.list.d/nodesource.list:
```
deb https://deb.nodesource.com/node_10.x stretch main
deb-src https://deb.nodesource.com/node_10.x stretch main
```

% apt-cache policy nodejs:
```
nodejs:
  Installed: 4.8.2~dfsg-1
  Candidate: 4.8.2~dfsg-1
  Version table:
     10.13.0-1nodesource1 500
        500 https://deb.nodesource.com/node_10.x stretch/main amd64 Packages
     8.11.2~dfsg-1 500
        500 http://deb.debian.org/debian sid/main amd64 Packages
 *** 4.8.2~dfsg-1 1000
       1000 http://deb.debian.org/debian stretch/main amd64 Packages
        100 /var/lib/dpkg/status
```

Great tipp! Looks like we found the culprit. What would you recommend to fix that priority?

**Comment by chrislea** [comment]
I'm pretty sure you already referenced it:
```
Pin: release a=stable
Pin-Priority: 1000
```
I believe that's putting anything in "stable", which is stretch in your case, to a priority of 1000, which overrides the default 500 that it's seeing on the package in our repository (and in the repository for sid, as you see). Hope this helps!

**Comment by GeraldWodni** [comment]
@chrislea thanks a lot for your help!

For any future googler I wanted to leave a detailed description on how to fix this issue.

## TL;DR: change your preferences to:
```
Package: *
Pin: release a=stable
Pin-Priority: 900 

Package: *
Pin: release o=Node Source
Pin-Priority: 950 

Package: *
Pin: release a=sid
Pin-Priority: 150 
```

## Long version:
I run my servers on debian stable, but occasionally need more up to date software from sid. In order to not interfere with `apt-get dist-upgrade` stable needs to be pinned above a Priority of 500. Doing this however will not allow you to install from nodesource (at least not without additional arguments to apt-get which is bad practice anyway).

So the solution is to increase the priority of packages from nodesource. In order to see a list of all repositories use `apt-cache policy`. This yields:
```
Package files:
 100 /var/lib/dpkg/status
     release a=now
 500 http://deb.debian.org/debian sid/main amd64 Packages
     release o=Debian,a=unstable,n=sid,l=Debian,c=main,b=amd64
     origin deb.debian.org
 950 https://deb.nodesource.com/node_10.x stretch/main amd64 Packages
     release o=Node Source,n=stretch,l=Node Source,c=main,b=amd64
     origin deb.nodesource.com
 900 http://deb.debian.org/debian stretch/non-free amd64 Packages
     release v=9.5,o=Debian,a=stable,n=stretch,l=Debian,c=non-free,b=amd64
     origin deb.debian.org
 ...
```
This also gives you the syntax for pinning, namely `o=Node Source` which you put into the preferences (see first listing).
